Add overview of CC problem handling (#33973)
gradle/gradle: platforms/core-configuration/configuration-cache/readme.md

# Configuration Cache

## Configuration Cache modes

CC can function in two modes:

- normal mode aka fail-on-problems
- warning mode aka lenient

Normal mode is how CC functions by default when enabled, ensuring safety and reliability.
Warning mode is an opt-in adoption and troubleshooting aid, meant to be used only temporarily.

## Problems and build failures

This section gives an overview of how CC problems interact with the build lifecycle,
when CC checks can result in build failures and how those are reported.

Whenever a problem is reported, it can be

1. **deferred** until the end of the build
2. **interrupting** the current scope of work

> 3. **suppressed** in an incompatible task

Author: Indeed, what we call a "suppressed problem" is more like an error downgraded to a warning -- still being reported, but no longer causes a failure. And suppressed silently is a sort of "weak warning" or even "info", because the performance is sub-optimal due to its presence (downgrade to vintage), but the user can't do anything about it.

I don't think we leak the "suppressed" terminology in the user-facing docs, so we can easily revise it later on.

---

I won't change the terminology here, but per your suggestion I added a paragraph in this section about all problems being included in the report.

# Configuration Cache

## Configuration Cache modes

CC can function in two modes:

- normal mode aka fail-on-problems
- warning mode aka lenient

> Normal mode is how CC functions in the majority of cases, ensuring safety and reliability.
> Warning mode is an adoption and troubleshooting aid, meant to be used only temporarily.

Reviewer: ```suggestion
Normal mode is how CC functions by default when enabled, ensuring safety and reliability.
Warning mode is an opt-in adoption and troubleshooting aid, meant to be used only temporarily.
```

## Problems and build failures

This section gives an overview of how CC problems interact with the build lifecycle,
when CC checks can result in build failures and how those are reported.

Whenever a problem is reported, it can be

1. **deferred** until the end of the build
2. **interrupting** the current scope of work
3. **suppressed** in an incompatible task
4. **suppressed silently** via graceful degradation

The way each problem is assigned severity depends on the CC mode and the context in which the problem appears.

> See `org.gradle.internal.cc.impl.problems.ProblemSeverity` usage in code.

Reviewer: 💅 All these `See X` notes would be better as links to the classes on GH, or to their javadocs.

### Deferred problems

The problems are *deferred* to maximize the number of reported problems,
providing the user with a more comprehensive overview of what needs fixing before CC can function successfully.

> This approach is similar to how compilers are showing all possible compilation problems, instead of only the first one encountered.

Reviewer: ```suggestion
This approach is similar to how compilers show all possible compilation problems, instead of only the first one encountered.
```

> Examples of deferred problems are serialization problems of tasks, which are expected.

Reviewer: ```suggestion
Examples of deferred problems include problems when serializing tasks, which are expected when the tasks are incompatible with CC.
```

For instance, when a task class or task action references/captures objects of forbidden types such as `Project` or `Configuration`.

> See `UnsupportedTypesCodecs.kt` for more examples.

Author: I rewrote this into a couple paragraphs
